composer/loggers/mosaicml_logger.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Log to the MosaicML platform."""
+
+from __future__ import annotations
+
+import collections.abc
+import logging
+import operator
+import os
+import time
+from functools import reduce
+from typing import TYPE_CHECKING, Any, Dict, Optional
+
+import torch
+
+from composer.loggers.logger import Logger
+from composer.loggers.logger_destination import LoggerDestination
+from composer.utils import MissingConditionalImportError, dist
+
+if TYPE_CHECKING:
+    from composer.core import State
+
+log = logging.getLogger(__name__)
+
+__all__ = ['MosaicMLLogger']
+
+RUN_NAME_ENV_VAR = 'RUN_NAME'
+
+
+class MosaicMLLogger(LoggerDestination):
+    """Log to the MosaicML platform.
+
+    Logs metrics to the MosaicML platform. Logging only happens on rank 0 every ``log_interval``
+    seconds to avoid performance issues.
+
+    Args:
+        log_interval (int, optional): Buffer log calls more frequent than ``log_interval`` seconds.
+                                      Defaults to 60.
+    """
+
+    def __init__(
+        self,
+        log_interval: int = 60,
+    ) -> None:
+        self.log_interval = log_interval
+        self._enabled = dist.get_global_rank() == 0
+        if self._enabled:
+            self.time_last_logged = 0
+            self.buffered_metadata: Dict[str, Any] = {}
+
+            try:
+                import mcli
+                del mcli
+            except ImportError as e:
+                raise MissingConditionalImportError(extra_deps_group='mcli',
+                                                    conda_package='mcli',
+                                                    conda_channel='conda-forge') from e
+            self.run_name = os.environ.get(RUN_NAME_ENV_VAR)
+            if self.run_name is not None:
+                log.info(f'Logging to mosaic run {self.run_name}')
+            else:
+                log.warning(f'Environment variable `{RUN_NAME_ENV_VAR}` not set, so MosaicMLLogger '
+                            'is disabled as it is unable to identify which run to log to.')
+                self._enabled = False
+
+    def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
+        self._log_metadata(hyperparameters)
+
+    def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None) -> None:
+        self._log_metadata(metrics)
+
+    def batch_end(self, state: State, logger: Logger) -> None:
```

Comment:
are you sure you want to flush on every batch? that doesn't limit the sending _that_ much

Reply:
It calls the flush function, but it won't flush unless the interval has passed

Reply:
oh right I kinda missed that. the intention of the flush is to _write even if the interval has not passed_. thats why  suggested running it at fit end (and maybe some other common checkpoints). Also would be good to flush on a crash if possible (i.e. in the close or post_close or whatever).